backend.native/tests/runtime/memory/cycle_detector.kt

```diff
@@ -0,0 +1,167 @@
+fun oneCycleWithArray() {
+    val array = arrayOf(AtomicReference<Any?>(null), AtomicReference<Any?>(null))
+    try {
+        array[0].value = Holder(array).freeze()
+        val cycles = GC.detectCycles()!!
+        assertEquals(1, cycles.size)
+        assertArrayEquals(arrayOf(array[0], array[0].value!!, array, array[0]), GC.findCycle(cycles[0])!!)
+    } finally {
+        array[0].value = null
+        array[1].value = null
+    }
+}
+
+@Test
+fun oneCycleWithLongChain() {
+    val atomic = AtomicReference<Any?>(null)
+    try {
+        val head = Holder(null)
+        var current = head
+        repeat(30) {
+            val next = Holder(null)
+            current.other = next
+            current = next
+        }
+        current.other = atomic
+        atomic.value = head.freeze()
+        val cycles = GC.detectCycles()!!
+        assertEquals(1, cycles.size)
+        val cycle = GC.findCycle(cycles[0])!!
+        assertEquals(33, cycle.size)
+    } finally {
+        atomic.value = null
+    }
+}
+
+@Test
+fun twoCycles() {
+    val atomic1 = AtomicReference<Any?>(null)
+    val atomic2 = AtomicReference<Any?>(null)
+    try {
+        atomic1.value = atomic2
+        atomic2.value = atomic1
+        val cycles = GC.detectCycles()!!
+        assertEquals(2, cycles.size)
+        assertArrayEquals(arrayOf(atomic2, atomic1, atomic2), GC.findCycle(cycles[0])!!)
+        assertArrayEquals(arrayOf(atomic1, atomic2, atomic1), GC.findCycle(cycles[1])!!)
+    } finally {
+        atomic1.value = null
+        atomic2.value = null
+    }
+}
+
+@Test
+fun twoCyclesWithHolder() {
+    val atomic1 = AtomicReference<Any?>(null)
+    val atomic2 = AtomicReference<Any?>(null)
+    try {
+        atomic1.value = atomic2
+        atomic2.value = Holder(atomic1).freeze()
+        val cycles = GC.detectCycles()!!
+        assertEquals(2, cycles.size)
+        assertArrayEquals(arrayOf(atomic2, atomic2.value!!, atomic1, atomic2), GC.findCycle(cycles[0])!!)
+        assertArrayEquals(arrayOf(atomic1, atomic2, atomic2.value!!, atomic1), GC.findCycle(cycles[1])!!)
+    } finally {
+        atomic1.value = null
+        atomic2.value = null
+    }
+}
+
+@Test
+fun threeSeparateCycles() {
+    val atomic1 = AtomicReference<Any?>(null)
+    val atomic2 = AtomicReference<Any?>(null)
+    val atomic3 = AtomicReference<Any?>(null)
+    try {
+        atomic1.value = atomic1
+        atomic2.value = Holder2(atomic1, atomic2).freeze()
+        atomic3.value = Holder2(atomic3, atomic1).freeze()
+        val cycles = GC.detectCycles()!!
+        assertEquals(3, cycles.size)
+        assertArrayEquals(arrayOf(atomic3, atomic3.value!!, atomic3), GC.findCycle(cycles[0])!!)
+        assertArrayEquals(arrayOf(atomic2, atomic2.value!!, atomic2), GC.findCycle(cycles[1])!!)
+        assertArrayEquals(arrayOf(atomic1, atomic1), GC.findCycle(cycles[2])!!)
+    } finally {
+        atomic1.value = null
+        atomic2.value = null
+        atomic3.value = null
+    }
+}
```

Comment:
Tests for `FreezableAtomicReference` would be useful.

Reply:
I mean the test that would check that `FreezableAtomicReference` aren't reported as leaking cycles.